lib/ecmarkdown.js

```diff
@@ -1,11 +1,357 @@
 'use strict';
-var parser = require('./generated-parser.js');
+
 var escapeHtml = require('escape-html');
+var beautify = require('./beautify.js');
 
-exports.list = function (ecmarkdown) {
-    return parser.parse(ecmarkdown.trim() + '\n', { startRule: 'list' });
+exports.document = function (ecmarkdown) {
+  return beautify(emitDocument(parse(ecmarkdown)));
 };
 
 exports.fragment = function (ecmarkdown) {
-    return parser.parse(escapeHtml(ecmarkdown.trim()), { startRule: 'fragment' });
+  return beautify(emitFragment(parse(escapeHtml(ecmarkdown))));
 };
+
+
+var lexemes = {
+  star: /\*/,
+  underscore: /_/,
+  tick: /`[^`\n]*`/,
+  pipe: /\|[^|\n]*\|/,
+  string: /"[^"\n]*"/,
+  tilde: /~/,
+  parabreak: /\n\n/,
+  linebreak: /\n/,
+  list: /[ \t]*\d+\. /,
+  whitespace: /[\s]+/,
+  comment: /<!--[\w\W]*-->/,
+  tag: /<[\/\w].*>/,
+  chars: /[^*_"~\s\n][^*_`"~<|\s\n]*/
+};
+
+var tokNames = Object.keys(lexemes);
+var tokRe = new RegExp(
+  tokNames.map(function (key) {
+    return '(^' + lexemes[key].toString().slice(1, -1) + ')';
+  })
+  .join('|')
+);
+
+var wrappers = {
+  star: 'emu-val',
+  underscore: 'var',
+  tick: 'code',
+  string: 'code',
+  tilde: 'emu-const'
+};
+
+function parse(emd) {
+  var t = new Tokenizer(emd);
+  var res = parseDocument();
+
+  return res;
+
+  function parseChars(inList, fmtStack) {
+    var node = { name: 'chars', contents: '' };
+    var tok = t.peek();
+
+    while (tok.name !== 'EOF' &&
+          (isChars(tok) || isWhitespace(tok) || isInvalidFormat(t.previous(), tok, t.peek(2), fmtStack))
+    ) {
+      t.next();
+      node.contents += tok.contents;
+      tok = t.peek();
+    }
+
+    return node;
+  }
+
+  function isInvalidFormat(prev, cur, next, fmtStack) {
```

Comment:
It's actually the tokenizer that is the main issue here. isSegmentNode is particularly disgusting as it doesn't take the node as a parameter, but rather gets the current token from the tokenizer. I was in the middle of this refactoring when I gave up and published this since it was taking me a while and I figured if I publish this code I'll stop doing complete rewrites (this is like the 4th iteration no joke).